sky/utils/kubernetes_utils.py

```diff
@@ -0,0 +1,468 @@
+
+    @classmethod
+    def get_label_key(cls) -> str:
+        return cls.LABEL_KEY
+
+    @classmethod
+    def get_label_value(cls, accelerator: str) -> str:
+        # For SkyPilot formatter, we use the accelerator str directly.
+        # See sky.utils.kubernetes.gpu_labeler.
+        return accelerator.lower()
+
+
+# LABEL_FORMATTER_REGISTRY stores the label formats SkyPilot will try to
+# discover the accelerator type from. The order of the list is important, as
+# it will be used to determine the priority of the label formats.
+LABEL_FORMATTER_REGISTRY = [SkyPilotLabelFormatter, GKELabelFormatter]
+
+
+def detect_gpu_label_formatter(
+) -> Tuple[Optional[GPULabelFormatter], List[Tuple[str, str]]]:
+    """Detects the GPU label formatter for the Kubernetes cluster
+
+    Returns:
+        GPULabelFormatter: The GPU label formatter for the cluster, if found.
+        List[Tuple[str, str]]: The set of labels and values across all nodes.
+    """
+    # Get all labels across all nodes
+    node_labels: List[Tuple[str, str]] = []
+    nodes = get_kubernetes_nodes()
+    for node in nodes:
+        node_labels.extend(node.metadata.labels.items())
+
+    label_formatter = None
+
+    # Check if the node labels contain any of the GPU label prefixes
+    for lf in LABEL_FORMATTER_REGISTRY:
+        label_key = lf.get_label_key()
+        for label, _ in node_labels:
+            if label.startswith(label_key):
+                label_formatter = lf()
+                break
+
+    return label_formatter, node_labels
+
+
+def detect_gpu_resource() -> Tuple[bool, Set[str]]:
+    """Checks if the Kubernetes cluster has nvidia.com/gpu resource.
+
+    If nvidia.com/gpu resource is missing, that typically means that the
+    Kubernetes cluster does not have GPUs or the nvidia GPU operator and/or
+    device drivers are not installed.
+
+    Returns:
+        bool: True if the cluster has nvidia.com/gpu resource, False otherwise.
+    """
+    # Get the set of resources across all nodes
+    cluster_resources: Set[str] = set()
+    nodes = get_kubernetes_nodes()
+    for node in nodes:
+        cluster_resources.update(node.status.allocatable.keys())
+    has_gpu = 'nvidia.com/gpu' in cluster_resources
+
+    return has_gpu, cluster_resources
+
+
+def get_kubernetes_nodes() -> List[Any]:
+    # TODO(romilb): Calling kube API can take between 10-100ms depending on
+    #  the control plane. Consider caching calls to this function (using
+    #  kubecontext hash as key).
+    try:
+        nodes = kubernetes.core_api().list_node(
+            _request_timeout=kubernetes.API_TIMEOUT).items
+    except kubernetes.max_retry_error():
+        raise exceptions.ResourcesUnavailableError(
+            'Timed out when trying to get node info from Kubernetes cluster. '
+            'Please check if the cluster is healthy and retry.') from None
+    return nodes
+
+
+def check_instance_fits(instance: str) -> Tuple[bool, Optional[str]]:
```

Comment:
Just a thought for discussion: maybe in the future, we can have a `kubernetes_catalog` module, where we use the similar APIs as the other cloud, but calling these online checking methods to check the availability. : )

Reply:
Ahh good point.. we started with having a hardcoded CSV catalog, then converged onto this design. Perhaps your suggestion of having a `kubernetes_catalog` is a good middle ground. I've left a TODO under `get_default_instance_type` for this :)